packages/toolkit/src/query/react/buildHooks.ts

```diff
@@ -502,7 +523,7 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
 
       const promiseRef = useRef<QueryActionCreatorResult<any>>()
 
-      useEffect(() => {
+      usePossiblyImmediateEffect((): void | undefined => {
```

Comment:
Since I'm not that familiar with RTK Query I'm just skimming this PR from my (SSR) perspective. I'm not sure how subscriptions work here (and don't have time to dig) so I'll go ahead and ask some questions. 😄

I'm not sure about converting this entire effect to be immediate. Specifically I notice that there is an effect cleanup calling `promiseRef.current?.unsubscribe()` on line 559. This is never called on the server, will that cause problems? (You mentioned hydration issues related to this in a Twitter-thread, is there anything else that might be problematic like memory leaks on the server?) I'm guessing `dispatch(initiate(...))` is what sets up that subscription? Is there a way to start a fetch and get a promise back without setting up a subscription and can that be used instead with `sideEffectsInRender`?

About the Suspense parts, who knows what that will look like when Suspense for data fetching is finished, but at a glance it looks like this could work in the current versions.

(Aside: I also noticed this effect will be called in every render in Suspense mode, but looks like it's basically a noop if nothing has changed so maybe that is fine? Might be unexpected and brittle if adding things to `usePossiblyImmediateEffect`s in the future though?)

Reply:
Having those subscriptions added in SSR might actually be helpful.

I'm thinking of a "hydrate" action (or other helper) that first restores the state from the server and then unsubscribes all subscriptions that were part of that "initial data". 
That would start all those "the last component unsubscribed" timers for cache cleanup while components could "re-claim" those cache entries themselves by subscribing to them.

In reality though, that might be quite a lot of actions flying around at once, so I'll have to experiment around with this a bit.

Reply:
As for the `usePossiblyImmediateEffects` being potentially brittle: I think we'll need to do something during render in every possible suspense scenario I can think of, so there is probably no way around that. I've just stuck this to the `useEffects` where it's necessary (also those without a cleanup) and hope that we can get away with it, at least until other patterns surface. It's also a marker that we have to handle these effects with extra care.

Reply:
Memory leaks on the server should probably not be a problem, if the Redux store is disposed of correctly. (If nothing unsubscribes, there will not even be any timers around for cache collection.)

It might be a good practice to clean up thoroughly in a SSR scenario by doing a `store.dispatch(api.util.resetApiState())` after the render is complete though, to make double sure really no timeouts etc. are sticking around. 

We should definitely add that in a SSR documentation.

Reply:
> That would start all those "the last component unsubscribed" timers for cache cleanup

Since I'm not familiar with what composes "a subscription" in RTK Query terms I'm just asking naïve questions here, but couldn't this be done for all queries in the store at hydration anyway? Does there need to be a previous subscription for this to work? (I get why that might make implementation easier though)

The reason I'm asking is that since components only render once in the same render on the server, subscriptions are usually a funky mental model conceptually there and most libraries avoid them for that reason to avoid having to deal with cleaning them up (something that often leaks into userland). To me it would seem unexpected if I inspected the store state after rendering on the server and I saw component subscriptions still lingering around. Doesn't mean avoiding subscriptions on the server is necessarily the right call here of course, just trying to provide context. 😄

> It's also a marker that we have to handle these effects with extra care

I think the naming is great and doing things in render is just how Suspense works so that part is great, the potentially unexpected part for me was just that the dependency array had no meaning in some cases. One should always code so effects works without the dep array anyway, so that's no big deal.

> Memory leaks on the server should probably not be a problem

Great! Just wanted to mention it as something to look out for.

Reply:
A "subscription" in this case is more like a "reference counter". It's the information "hey, there is a component out there, displaying that data, please don't remove it from the cache". (There can be more information attached like "this component requests for the data to be polled", but I feel that's not of major importance here). 
So you might be reading more into the word that it actually means :slightly_smiling_face: 

I agree that we could do stuff like "never subscribe in the first place", but if we go this route of restoring a "live cache snapshot", it would also enable users of local storage persistence to rehydrate. 
While I personally think that most times, cache data shouldn't be persisted and if you want to have that kind of cache, just set the right HTTP headers and let the browser handle it, I'm sure there are some edge cases around where data has to be persisted - and having a rehydration tool around for that would also be handy.

It might also come down to removing subscriptions before restoring data, if it is contained in the snapshot. I guess I'll explore many options there. That is all a bit muddy before I get to it :sweat_smile:

Reply:
Ah, I was starting to suspect that a "subscription" in RTK Query was a way to _track_ subscriptions/dependents rather than actually _subscribing_ since that seems like something you would do through regular Redux means. In React Query there's the concept of "observers" which seems very similar. This does make things fall into place, thanks!

In fact, I'm pretty sure we've at some point had a bug in React Query where data fetched by prefetch was in some cases prematurely thrown away because it had no observers or something like that, so providing a way of fetching data without counting that as an "observer" also has tradeoffs.

I guess a different way to put it is, when you support imperative fetching with cache eviction, you either:

* Register that as an "dependent" and require the developer to later "unregister" that manually (the hydration tool you talk about)
* Don't register a "dependent", but now the data is immediately stale/scheduled for cache eviction, or you place it in the cache for infinite time, or you need to find some other way to deal with it

I guess it boils down to supporting imperative fetching in libraries where "query lifetime" is a central concept is kind of a mismatch that requires working around one way or the other. 😄